aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs

```diff
@@ -90,13 +83,17 @@ impl<'a, S: 'a + StateView + Sync> ExecutorTask for AptosExecutorTask<'a, S> {
                     ExecutionStatus::DelayedFieldsCodeInvariantError(
                         vm_status.message().cloned().unwrap_or_default(),
                     )
-                } else if AptosVM::should_restart_execution(&vm_output) {
+                } else if AptosVM::should_restart_execution(vm_output.change_set()) {
                     speculative_info!(
                         &log_context,
                         "Reconfiguration occurred: restart required".into()
                     );
                     ExecutionStatus::SkipRest(AptosTransactionOutput::new(vm_output))
                 } else {
+                    assert!(
+                        Self::is_transaction_dynamic_change_set_capable(txn),
+                        "DirectWriteSet should always create SkipRest transaction, validate_waypoint_change_set provides this guarantee"
```

Comment:
Currently, the `validate_waypoint_change_set` function verifies that change set contains new block event and new epoch event.

Does this means this transaction has to be the first transaction in the first block of an epoch? 

This doesn't directly verify that the waypoint transaction should create SkipRest as output right?

Reply:
new epoch event will be caught by the above else-if, and that will create SkipRest, if I follow this correctly

Reply:
yup, epoch event creates SkipRest